Document ES module conventions

robertknight opened the pull request:
Document our existing conventions around ES module usage:

 - Naming of ES modules
 - Sorting and grouping of `import`s
 - Use of named and default exports

Additionally add a note at the top to highlight the convention of
putting the rationale for choices in italics.

robertknight committed:
Document ES module conventions

Document our conventions around ES module usage:

 - Naming of ES modules
 - Sorting and grouping of `import`s
 - Use of named and default exports

Additionally add a note at the top to highlight the convention of
putting the rationale for choices in italics.

robertknight committed:
Specify language for several code blocks

esanzgar commented on docs/js-guide.md:
```diff
@@ -33,17 +36,58 @@ Projects may use any ES language features which meet all of the following criter
 
+### Module naming
+
+Modules should use lowercase, hyphen-separated names (`helpful-module.js`).
```
Must the module's names have two parts (separated by hyphen)?

robertknight replied:
Only if it has multiple words. If the module name is just one word then it doesn't need hyphens.

esanzgar replied:
Thanks for answering my question. I thought I needed to include a second word always.

esanzgar reviewed (approved):
👍🏼 

Thanks, very useful for me.

lyzadanger commented on docs/js-guide.md:
```diff
@@ -33,17 +36,58 @@ Projects may use any ES language features which meet all of the following criter
+Imports should be placed at the top of a module, grouped into sections by their
+relation to the current project:
+
+- Platform (node, browser) imports
```
Perhaps this could be a numbered list since this is ordered?

lyzadanger commented on docs/js-guide.md:
```diff
@@ -33,17 +36,58 @@ Projects may use any ES language features which meet all of the following criter
+- Imports from other directories in the current package
+- Imports from the same directory in the current package
+
+Within each group import lines should be sorted alphabetically by module path.
```
I have a lingering minor confusion on this: is this alphabetization by the entire path as a single string? I struggle with this, oddly! Here's a good example from a client module:

```
import { useStoreProxy } from '../store/use-store';
import { withServices } from '../service-context';
import VersionData from '../helpers/version-data';
```

Should this be:

```
import VersionData from '../helpers/version-data';
import { withServices } from '../service-context';
import { useStoreProxy } from '../store/use-store';
```

without regard to path depth?

p.s. I think I'm the only one confused by this!

robertknight replied:
Entire path as a single string, so matching your second example. I'd hope we can automate this in future, though I haven't got a working setup yet.

lyzadanger commented on docs/js-guide.md:
```diff
@@ -3,9 +3,12 @@
 This document explains the tools and conventions that we use to maintain
 quality and consistency of our JavaScript code.
 
+The rationales for various conventions are documented briefly _in italics_.
```
This makes for a fair amount of text in italic...I like the idea of having a "do this"..."because this" structure, though, that's good!

lyzadanger commented on docs/js-guide.md:
```diff
@@ -33,17 +36,58 @@ Projects may use any ES language features which meet all of the following criter
+
+### Default exports
+
+Modules should always use named exports rather than default exports.
```
I wonder if we might avoid using the word _always_ in a convention/rule?

robertknight replied:
I've reworded this sentence in a way that avoids "always" and is hopefully clearer.

lyzadanger reviewed (approved):
Thanks for these great updates!

robertknight committed:
Apply PR feedback

 - Reword "Default exports" section for clarity
 - Use a numbered list for the import order list
